CenterForOpenScience/exp-addons
# Counterbalancing randomizer should act on previous session

Refs: https://openscience.atlassian.net/browse/LEI-215
Companion to: https://github.com/CenterForOpenScience/lookit/pull/60/files

## Purpose
The pref-phys-pilot randomizer should take previous sessions into account when deciding what videos to show. 

Previously, it did not, because it was looking for the previous frame under a key that did not exist. See JIRA ticket for notes on expected behavior.

## Summary of changes
- Get unit tests working
- Fix pref-phys-pilot randomizer to correctly find the prev session conditions
- Add unit tests around the expected behavior so we know if it breaks in the future

## Testing notes
1. Make sure your account has participated less than 50 times (else you will get stuck on the first page of results and every single new session will be based on the last item of the first page- after 50 tries your payload will always be the same!)
2. Go to the studies page and click "participate in a session"
3. In the network inspector, find the last PATCH request to send session data to the server, and inspect the "attributes.conditions" section of the payload as below
4. Without completing the experiment, go to the studies page and participate again. Compare this payload to the previous one.

## Sample results
```
NEW CHILD:

First run (no lastSession description available)
	.	Request URL:Request URL:https://hostname/v1/id/documents/namespace.collection123.recordABB
	.	
	.	{7-pref-phys-videos: {whichObjects: [0, 5, 2, 5], NPERTYPE: 6, startType: 2, showStay: 1}}



Next run:
lastFrameConditions: 
"{
    "whichObjects": [
        0,
        5,
        2,
        5
    ],
    "NPERTYPE": 6,
    "startType": 2,
    "showStay": 1
}"

	.	Request URL:https://hostname/v1/id/documents/namespace.collection123.recordABC
	.	
	.	conditions:{7-pref-phys-videos: {whichObjects: [1, 0, 3, 0], NPERTYPE: 6, startType: 3, showStay: 0}}
```
